### PR TITLE
feat(Dataset): migrate dataset endpoints to API v4

### PR DESCRIPTION
**Pull request by abdimo101**
## Description
This PR replaces dataset related API calls from V3 to V4, updating DTOs, effects, reducers, components and various tests to align with V4. Also added a utils file inside shared-condition folder that handles the scientific metadata query logic to be used for the findAllV4 endpoint.

Skipped test file:
- datasets-metadata.cy.js

Reason for the test skips are because backend does not yet support:
- Scientific metadata key encoding/decoding (PR: https://github.com/SciCatProject/backend/pull/2765)

## Motivation
Background on use case, changes needed

## Fixes:
Please provide a list of the fixes implemented in this PR

* Items added

## Changes:
Please provide a list of the changes implemented by this PR

* changes made

## Tests included
- [ ] Included for each change/fix?
- [ ] Passing? (Merge will not be approved unless this is checked) 

## Documentation
- [ ] swagger documentation updated \[required\]
- [ ] official documentation updated \[nice-to-have\]

### official documentation info
If you have updated the official documentation, please provide PR # and URL of the pages where the updates are included

## Backend version
- [ ] Does it require a specific version of the backend
- which version of the backend is required:

## Summary by Sourcery

Migrate dataset operations and frontend models from API V3 to V4 while supporting V4 query semantics and multi-valued dataset relationships.

New Features:
- Support dataset queries through the V4 API, including authenticated and public access paths.
- Support multiple proposals, samples, and instruments associated with a dataset.
- Add scientific metadata condition-to-query conversion with SI unit handling and decimal range values.

Bug Fixes:
- Correct dataset creation and filtering payloads to use V4 field and query formats.
- Improve dataset detail loading and state clearing when resolving multiple related entities.

Enhancements:
- Replace legacy dataset DTOs and state types with V4-compatible dataset models across dataset views, effects, reducers, and shared services.
- Update dataset-related Cypress commands, fixtures, intercepts, and coverage for V4 behavior.
- Simplify dataset detail tab and data-loading behavior around the current dataset.

Tests:
- Update unit and end-to-end tests for V4 services, DTOs, payloads, and multi-valued relationships.
- Skip scientific metadata metadata-key Cypress coverage until backend encoding and decoding support is available.

### CODE REVIEW

**Review by Junjiequan — DISMISSED** [review]
The v4 datasets lookup feature integration is missing here, which was the main driver for the v3 to v4 migration. Since this lookup capability is the core reason we moved to v4, we should make sure it's included before this is considered complete, otherwise the migration doesn't deliver its primary benefit.

File: src/app/datasets/dataset-lifecycle/dataset-lifecycle.component.ts [hunk 1]
```diff
@@ -47,7 +45,7 @@ export type HistoryWithProperties = HistoryClass & { [key: string]: unknown };
 export class DatasetLifecycleComponent implements OnInit, OnChanges {
   appConfig = this.appConfigService.getConfig();
 
-  dataset: OutputDatasetObsoleteDto | undefined;
+  dataset: CurrentDataset | undefined;
```
Comment (nitrosx) [inline src/app/datasets/dataset-lifecycle/dataset-lifecycle.component.ts:48]: is `CurrentDataset` the correct DTO, correct naming?
Please rename to `CurrentDatasetDTO` if it is indeed a DTO

File: src/app/datasets/reduce/reduce.component.ts
Comment (nitrosx) [inline src/app/datasets/reduce/reduce.component.ts:1]: `CurrentDataset`?

File: src/app/logbooks/logbooks-dashboard/logbooks-dashboard.component.ts [hunk 2]
```diff
@@ -52,7 +50,7 @@ export class LogbooksDashboardComponent
 {
   vm$ = this.store.select(selectLogbooksDashboardPageViewModel);
 
-  dataset: OutputDatasetObsoleteDto | undefined = undefined;
+  dataset: CurrentDataset | undefined = undefined;
```
Comment (nitrosx) [inline src/app/logbooks/logbooks-dashboard/logbooks-dashboard.component.ts:53]: `CurrentDataset`?

File: src/app/shared/services/ownership.service.ts [hunk 3]
```diff
@@ -3,19 +3,19 @@ import { Router } from "@angular/router";
...
+import { CurrentDataset } from "state-management/state/datasets.store";
 
 @Injectable({
   providedIn: "root",
 })
 export class OwnershipService {
   public isOwner = false;
   checkDatasetAccess(
-    dataset: OutputDatasetObsoleteDto | undefined,
+    dataset: CurrentDataset | undefined,
```
Comment (nitrosx) [inline src/app/shared/services/ownership.service.ts:18]: `CurrentDataset`?

File: src/app/state-management/state/datasets.store.ts [hunk 4]
```diff
@@ -16,11 +25,21 @@ export interface Pagination {
...
+};
+
 export interface DatasetState {
-  datasets: OutputDatasetObsoleteDto[];
-  selectedSets: OutputDatasetObsoleteDto[];
-  currentSet: OutputDatasetObsoleteDto | undefined;
-  relatedDatasets: OutputDatasetObsoleteDto[];
+  datasets: PartialOutputDatasetDto[];
+  selectedSets: PartialOutputDatasetDto[];
+  currentSet: CurrentDataset | undefined;
```
Comment (nitrosx) [inline src/app/state-management/state/datasets.store.ts:41]: `CurrerntDataset`?

**Review by nitrosx — CHANGES_REQUESTED** [review]

**Comment by nitrosx** [comment]
# Code Review Report: change-dataset-endpoints-to-v4 Branch

Generated by Mistral Vibe.
Co-Authored-By: Mistral Vibe <vibe@mistral.ai>

---

## Overview

1. **Verify logic correctness**: The changes successfully migrate from v3 to v4 API endpoints. The logic is generally correct with proper handling of the new data structures and API response formats. The migration maintains backward compatibility in several places.

2. **Check if all edge cases are handled**: Most edge cases are handled, particularly:
   - Legacy fallback for attachment IDs (`attachmentIdWithLegacyFallback`)
   - Null checks for optional fields
   - Empty array handling in filters
   However, there are some gaps in error handling for failed API calls.

3. **Summarize what the code touched by the changes does**: This branch migrates the Scicat frontend application from using v3 REST API endpoints to v4 endpoints for datasets and attachments. It updates data models, API service calls, state management, and UI components to work with the new API structure which uses MongoDB-style queries instead of Elasticsearch-style queries.

4. **Assess whether the changes make sense**: The changes are well-structured and necessary for API version compatibility. The migration is comprehensive and touches all necessary layers (actions, effects, selectors, components). The approach of maintaining backward compatibility where needed (like the attachment ID fallback) is pragmatic.

5. **Identify any unreachable code**: The removed code for datablocks and origdatablocks fetching effects is no longer reachable as these actions are no longer dispatched. The elasticsearch-related commands in cypress/support/commands.js have been removed as they are no longer used.

**Commits responsible**: Main changes introduced in commits `fff5bae86` (v4 apis for attachment and dataset) and `645c34c52` (dynamic dataset detail to adopt dataset api v4), with subsequent fixes in `10bf4bd48`, `55b5f9bc2`, `13054ac11`, and `bf8e7952f`.

---

---

## Code Changes

### Major Changes Summary

| Area | Change Description | Files Modified |
|------|-------------------|----------------|
| **API Migration** | Switched from v3 endpoints (`datasetsControllerFindByIdV3`, `datasetsControllerFullqueryV3`) to v4 endpoints (`datasetsV4ControllerFindByIdV4`, `datasetsV4ControllerFindAllV4`) | datasets.effects.ts, datasets.actions.ts |
| **Data Models** | Replaced `OutputDatasetObsoleteDto` and `OutputAttachmentV3Dto` with `PartialOutputDatasetDto` and `OutputAttachmentV4Dto` | datasets.actions.ts, datasets.store.ts, various components |
| **Query Format** | Changed from Elasticsearch query format to MongoDB query format (`{where: {}, limits: {}}`) | datasets.effects.ts, datasets.selectors.ts |
| **Attachment Relationships** | V4 attachments use `relationships` array instead of direct `datasetId` field | datasets.effects.ts, file-uploader.component.ts |
| **Date Filtering** | Changed from `{begin: string, end: string}` to `{$gte: {$date: string}, $lte: {$date: string}}` | datasets.selectors.ts, datasets-filter.component.ts |
| **Removals** | Removed datablocks/origdatablocks fetching effects, elasticsearch commands | datasets.effects.ts, cypress/support/commands.js |

### Assessment

**Necessary**: Yes, the changes are necessary for API compatibility. The v3 endpoints are being deprecated in favor of v4, which provides a more standardized MongoDB-style query interface.

**Architecture**: The changes follow a clean separation of concerns:
- Actions define the intent
- Effects handle the API calls
- Selectors transform the data
- Components consume the data

The migration maintains the existing NgRx state management pattern while adapting to the new API.

**Improvement Needed**:
1. **Error Handling**: Some API calls lack comprehensive error handling. The `catchError` operators should include error logging for debugging purposes.
2. **Type Safety**: While types have been updated, there could be more specific type guards for handling the union types (v3 vs v4 attachments).
3. **Documentation**: Add comments explaining the v3-to-v4 migration in key files to help future maintainers understand the transition.
4. **Testing**: Ensure all new query formats are thoroughly tested, especially edge cases with empty filters and null values.

### Verdict

**Final Judgment**: The changes are **well-executed and necessary**. The migration from v3 to v4 APIs is comprehensive and maintains backward compatibility where needed. The code follows good practices and the changes make sense architecturally. The few areas for improvement (error handling, type safety) are minor compared to the overall quality of the migration.

---

---

## Security Review

1. **Are there any potential injection vulnerabilities?**
   - **No direct injection vulnerabilities found**. The code properly uses parameterized API calls through the SDK services rather than string concatenation.
   - The query parameters are properly JSON.stringify'd before being sent to the API, which prevents injection.
   - Commit `9bd5ffa17` explicitly removed elasticsearch functions from commands, which could have been a vector for query injection.

2. **Does this code expose any sensitive user data?**
   - **No new data exposure**. The changes maintain the same data access patterns as before.
   - The v4 API endpoints have the same authentication requirements as v3.
   - User access groups are still properly checked before displaying restricted data.

3. **Are there instances of insecure API usage?**
   - **One concern**: In `datasets.effects.ts`, the `removeAttachment$` effect has a comment noting a workaround for a backend bug: `// NOTE: the second argument should be removed on new backend release after v5.2.1`. This uses an empty string as a workaround, which could be problematic if the backend behavior changes.
   - The token handling in cypress tests was improved - commit `10bf4bd48` and others standardized the use of `getHeader(token)` for authentication headers.

4. **Could this code lead to an authentication bypass?**
   - **No authentication bypass vulnerabilities identified**. The authentication flow remains unchanged.
   - The code properly differentiates between authenticated (v4) and public (publicV4) endpoints based on user login status.
   - User token is still required for write operations.

**Commits responsible**:
- `fff5bae86`: Main v4 API migration
- `645c34c52`: Dataset detail adaptation
- `9bd5ffa17`: Removed elasticsearch functions
- `10bf4bd48`: Added attachment ID fallback handling

---

---

## Test Coverage

### Current Test Coverage

1. **Unit Tests**:
   - `datasets.actions.spec.ts`: Tests exist for actions, updated to use new types
   - `datasets.effects.spec.ts`: Tests updated to mock v4 services
   - `datasets.reducer.spec.ts`: Tests updated for new state structure
   - `datasets.selectors.spec.ts`: Tests updated for new selector logic

2. **E2E Tests**:
   - `datasets-attachment.cy.js`: Updated to use v4 API endpoints (`lbBaseUrlV4`)
   - `datasets-general.cy.js`: Updated to use v4 endpoints
   - `datasets-keyword.cy.js`: Updated to use v4 endpoints
   - `datasets-metadata.cy.js`: Updated to use v4 endpoints
   - `datasets-public.cy.js`: Updated to use v4 endpoints
   - `main-page.cy.js`: Updated accordingly

3. **Component Tests**:
   - `dataset-detail-dynamic.component.spec.ts`: Extensively updated to work with new data structures
   - Tests for filters and date ranges have been updated

### Assessment

**Strengths**:
- Comprehensive test updates across the board
- Both unit and E2E tests cover the new API endpoints
- Tests validate the new query formats and data structures

**Gaps Identified**:
1. **Missing Tests for Edge Cases**:
   - No explicit tests for the `attachmentIdWithLegacyFallback` function in file-uploader component
   - No tests for the `toLegacyDateRange` conversion function
   - Limited tests for error scenarios in API calls

2. **Date Range Filter Testing**:
   - The new MongoDB-style date range filters (`$gte`, `$lte`) need explicit tests to verify they produce the correct backend queries

3. **Public vs Authenticated Endpoints**:
   - Need tests to verify that public users cannot access authenticated-only data through the v4 endpoints

4. **Relationship Handling**:
   - The new relationship-based attachment handling should have integration tests

### Suggestions for Improvement

1. **Add Unit Tests**:
   ```typescript
   // Test for attachmentIdWithLegacyFallback
   describe('attachmentIdWithLegacyFallback', () => {
     it('should return aid for v4 attachments', () => {
       const v4Attachment = { aid: '123', id: '456' };
       expect(component.attachmentIdWithLegacyFallback(v4Attachment)).toBe('123');
     });
     it('should return id for v3 attachments', () => {
       const v3Attachment = { id: '789' };
       expect(component.attachmentIdWithLegacyFallback(v3Attachment)).toBe('789');
     });
   });
   ```

2. **Add Integration Tests**:
   - Test the complete flow from UI action to v4 API call
   - Verify that authenticated vs public API endpoints are correctly selected

3. **Add Error Handling Tests**:
   - Test network failures
   - Test malformed responses
   - Test authentication failures

**Relevant commits**: `5da047ce3` (fix cypress test), `b438f8630` (fix unit test), `c4e2f5d45` (fix unit tests and eslint), `cb26ae533` (fix e2e)

---

---

## Security Examples

### Potential Injection Vulnerability Example

**Scenario**: If query parameters were concatenated directly into API URLs without proper encoding.

**Code that avoids this** (from datasets.effects.ts):
```typescript
// GOOD: Uses JSON.stringify which properly escapes the query
this.datasetsV4Service.datasetsV4ControllerFindAllV4(
  JSON.stringify(filter),
)
```

**Vulnerable alternative** (not present in codebase):
```typescript
// BAD: Direct string concatenation
const query = `?filter=${userInput}`;
return this.http.get(`/api/datasets${query}`);
```

### Authentication Bypass Example

**Scenario**: If the code didn't properly check user authentication before making authenticated API calls.

**Code that prevents this** (from datasets.effects.ts):
```typescript
// GOOD: Properly differentiates between authenticated and public endpoints
const apiCall$ = user
  ? this.datasetsV4Service.datasetsV4ControllerFindAllV4(JSON.stringify(filter))
  : this.datasetsPublicV4Service.datasetsPublicV4ControllerFindAllPublicV4(JSON.stringify(filter));
```

---

---

## Testing for Security Use Cases

### Testing for Injection Vulnerabilities

**Test Case**: Verify that user input in filters is properly escaped

**Test Code**:
```typescript
it('should properly escape special characters in query filters', () => {
  const maliciousInput = '{ $ne: "" }'; // MongoDB query injection attempt
  // This should be properly escaped by JSON.stringify
  const filter = { datasetName: maliciousInput };
  const query = JSON.stringify(filter);
  // Verify that the query is properly stringified, not evaluated
  expect(query).toBe('{"datasetName":"{ $ne: \"\" }"}');
});
```

**File and Line**: `src/app/state-management/effects/datasets.effects.ts:75-80`

### Testing for Data Exposure

**Test Case**: Verify that sensitive fields are not exposed in public endpoints

**Test Code**:
```typescript
it('should not return sensitive data for public users', () => {
  // Mock public API response
  const publicResponse = { /* dataset data without sensitive fields */ };
  const authResponse = { /* dataset data with sensitive fields */ };

  // Verify that public endpoint is used when user is not authenticated
  const user = null;
  const filter = {};
  const apiCall$ = user
    ? of(authResponse)
    : of(publicResponse);

  apiCall$.subscribe(result => {
    expect(result).toEqual(publicResponse);
    expect(result).not.toContain('sensitiveField');
  });
});
```

**File and Line**: `src/app/state-management/effects/datasets.effects.ts:70-85`

### Testing for Authentication Requirements

**Test Case**: Verify that write operations require authentication

**Test Code**:
```typescript
it('should fail to create attachment without authentication', () => {
  // Attempt to create attachment without being logged in
  const attachment = { datasetId: '123', caption: 'test' };
  const apiCall$ = this.attachmentsV4Service.attachmentsV4ControllerCreateAttachmentV4(attachment);

  // This should return a 401 or similar error
  apiCall$.subscribe({
    next: () => fail('Should have failed without authentication'),
    error: (err) => expect(err.status).toBe(401)
  });
});
```

**File and Line**: `src/app/state-management/effects/datasets.effects.ts:352-370`

---

---

## Summary

### Key Findings

1. **Migration Quality**: The v3 to v4 API migration is **comprehensive and well-executed**. All layers of the application (actions, effects, selectors, components) have been properly updated.

2. **Security**: **No new security vulnerabilities** were introduced. The code properly:
   - Uses parameterized API calls through the SDK
   - Differentiates between authenticated and public endpoints
   - Maintains proper authentication checks
   - Removed potentially vulnerable elasticsearch functions

3. **Backward Compatibility**: The migration includes **pragmatic fallback mechanisms** (e.g., `attachmentIdWithLegacyFallback`) to handle the transition period where both v3 and v4 data structures may exist.

4. **Testing**: Test coverage is **generally good** but has some gaps in:
   - Error handling scenarios
   - Edge cases for data conversion functions
   - Complete integration tests for the new query formats

### Recommendations

1. **Immediate Actions**:
   - Monitor the workaround in `removeAttachment$` effect (empty string parameter) and remove it once the backend bug is fixed
   - Add explicit error logging in `catchError` operators for easier debugging

2. **Short-term Improvements**:
   - Add unit tests for `attachmentIdWithLegacyFallback` and date range conversion functions
   - Add integration tests for the complete v4 API flow
   - Add tests for error scenarios and edge cases

3. **Long-term Considerations**:
   - Plan for complete removal of v3 compatibility code once migration is complete
   - Document the v3-to-v4 changes for future maintainers
   - Consider creating a migration guide for other developers

### Overall Assessment

**Verdict**: The changes in the `change-dataset-endpoints-to-v4` branch are **production-ready**. The migration is thorough, maintains backward compatibility where necessary, and doesn't introduce new security vulnerabilities. The few identified gaps are minor and can be addressed in follow-up work.

The branch successfully achieves its goal of migrating the Scicat frontend to use v4 API endpoints for datasets and attachments while maintaining application functionality and security.

**Comment by Junjiequan** [comment]
@abdellah257  Could you test if the api migration breaks anything on your side. I removed instrumentId to instrumentName conversion function that was implemented from your side on dataset table and replaced them from v4 `lookup`.

File: src/app/datasets/dataset-lifecycle/dataset-lifecycle.component.ts (hunk 1 above)
Comment (Junjiequan) [reply src/app/datasets/dataset-lifecycle/dataset-lifecycle.component.ts:48]: currentDataset is frontend specific type, which can include lookup fields, `proposals`, 'instruments' etc. So it should not be mixed with outputDatasetDto from backend

File: src/app/datasets/reduce/reduce.component.ts
Comment (Junjiequan) [reply src/app/datasets/reduce/reduce.component.ts:1]: same as above

File: src/app/logbooks/logbooks-dashboard/logbooks-dashboard.component.ts (hunk 2 above)
Comment (Junjiequan) [reply src/app/logbooks/logbooks-dashboard/logbooks-dashboard.component.ts:53]: same as above

File: src/app/shared/services/ownership.service.ts (hunk 3 above)
Comment (Junjiequan) [reply src/app/shared/services/ownership.service.ts:18]: same as above

File: src/app/state-management/state/datasets.store.ts (hunk 4 above)
Comment (Junjiequan) [reply src/app/state-management/state/datasets.store.ts:41]: same as above

File: src/app/datasets/dataset-detail/dataset-detail-dynamic/dataset-detail-dynamic.component.html [hunk 5]
```diff
@@ -79,84 +83,92 @@
...
+                        let-isEmpty="isEmpty"
                       >
-                        <!-- InternalLink Field-->
-                        <ng-container *ngIf="fieldType === 'internalLink'">
-                          <ng-container *ngFor="let value of fieldValue">
-                            <a
-                              (click)="
-                                onClickInternalLink(
-                                  field.source,
-                                  getInternalLinkValue(dataset, field.source)
```
Comment (abdellah257) [inline src/app/datasets/dataset-detail/dataset-detail-dynamic/dataset-detail-dynamic.component.html]: In this case the `internalLink` behaviour would change to show the value instead if the `id`, not the use case we are using it for,  I suggest we keep it for id and provide another option for value link maybe ?

File: src/app/datasets/dataset-detail/dataset-detail-dynamic/dataset-detail-dynamic.component.ts [hunk 6]
```diff
@@ -31,23 +35,35 @@ import {
...
-  selectProfile,
-  (profile) => profile?.accessGroups || [],
-);
+import { CurrentDataset } from "state-management/state/datasets.store";
+
+const RELATION_CONFIG: Record<
+  string,
+  { lookupField: string; idField: string; labelField: string }
+> = {
+  proposalIds: {
```
Comment (abdellah257) [inline src/app/datasets/dataset-detail/dataset-detail-dynamic/dataset-detail-dynamic.component.ts:51]: In this case the `internalLink` behaviour would change to show the `title` instead if the `id` for proposals , not the use case we are using it for currently (id),  I suggest we keep it for id for all fields and provide configuration if a change is needed perhaps ?

File: src/app/datasets/dataset-details-dashboard/dataset-details-dashboard.component.ts [hunk 7]
```diff
@@ -123,193 +174,63 @@ export class DatasetDetailsDashboardComponent
...
     this.subscriptions.push(
       this.route.params
         .pipe(map((params) => params["id"]))
         .subscribe((id: string) => {
           if (id) {
-            this.resetTabs();
-            // Fetch dataset details
-            this.store.dispatch(fetchDatasetAction({ pid: id }));
-            this.fetchDataActions[TAB.details].loaded = true;
+            this.store.dispatch(
```
Comment (abdellah257) [inline src/app/datasets/dataset-details-dashboard/dataset-details-dashboard.component.ts:188]: The issue here is that all information related to a dataset is loaded on a single request unlike previously, this provides good code refactoring, however it becomes an issue if the dataset has a large number of files/OrigDataBlocks, the user will have to fetch these whether he is interested in checking the files at the datafiles TAB or not.

Also fetching all origdatablocks is not needed when dynamic datafiles table will be introduced.

**Review by abdellah257 — CHANGES_REQUESTED** [review]
Thanks for the changes !
The changes work fine for my side, I have mentioned the 2 issues I have. However I noticed that the text search doesn't work with previous backends (idk if it's related to this PR but I had to update the backend ) maybe we should specify which backend supports this change. otherwise I got this error
```json
{
    "message":"Property where.$text.$search should not exist in the filter object",
    "error":"Bad Request",
    "statusCode":400
}
```

File: src/app/datasets/dataset-detail/dataset-detail-dynamic/dataset-detail-dynamic.component.ts (hunk 6 above)
Comment (Junjiequan) [reply src/app/datasets/dataset-detail/dataset-detail-dynamic/dataset-detail-dynamic.component.ts:51]: That's a good point. I just blindly assumed that showing title would be ideal compare to showing ids. 
Yes, we should make it configurable, allowing site admin to decide whether to show title or id

File: src/app/datasets/dataset-details-dashboard/dataset-details-dashboard.component.ts (hunk 7 above)
Comment (Junjiequan) [reply src/app/datasets/dataset-details-dashboard/dataset-details-dashboard.component.ts:188]: That makes sense, maybe should we make the query configurable as well? so that different facilities can decide what fields to fetch. 

Regarding `$text.$search` query error, yes, we need new SDK for that, it's for backward compatibility with v3 fullquery behavior. The PR got merged recently but yet released

Comment (Junjiequan) [reply src/app/datasets/dataset-details-dashboard/dataset-details-dashboard.component.ts]: all the comments are addressed in [8cd2975](https://github.com/SciCatProject/frontend/pull/2380/commits/8cd2975ab393e6ef9a001585c9d05042038efe06)

I have added 2 new configurations:
1. `datasetDetailsTabsInclude` controlls additional fields to fetch using v4 findall lookup. default as below:

```js
  "datasetDetailsTabsInclude": {
    "details": [],
    "jsonScientificMetadata": [],
    "datafiles": ["origdatablocks"],
    "relatedDatasets": [],
    "relationships": [],
    "reduce": [],
    "logbook": [],
    "attachments": ["attachments"],
    "lifecycle": [],
    "admin": ["datablocks"]
  },
```
2. internalLinkLabel controls which field to show as label, default/fallback is ID if not provided or field do not exist

```js
      {
        "type": "regular",
        "label": "Related Documents",
        "order": 4,
        "col": 4,
        "row": 1,
        "fields": [
          {
            "element": "internalLink",
            "source": "instrumentIds",
            "internalLinkLabel": "id",
            "order": 1
          },
```

Comment (abdellah257) [reply src/app/datasets/dataset-details-dashboard/dataset-details-dashboard.component.ts]: Thank you @Junjiequan for the changes, I have tested them and they look good !

**Review by abdellah257 — APPROVED** [review]

**Comment by abdellah257** [comment]
I would like to mention that for published data, now it will be mandatory to update the isPublished status of all the other fields (origdatablocks, attachement, ...), unlike previousely !
It may be worth introducing (not related to this) a {pid}/publish endpoint with include options to avoid patching every element seperatly.
